### DISCUSSION TITLE
Ustream AttributeError: 'UHSStream' object has no attribute 'args'

### DISCUSSION THREAD

**Discussion by dsrushton** [streamlink/streamlink#3892] [Q&A] [answered]
Hi y'all, beginner here. 

Trying to get a ustream livefeed to record keep getting the error:

File "c:\Users\dsrus\Desktop\School\464\Visual Studio\Hello\Hello.py", line 110, in <module>
    dl_stream(videoURL, tempFile, 3)
  File "c:\Users\dsrus\Desktop\School\464\Visual Studio\Hello\Hello.py", line 55, in dl_stream
    stream_segment = get_stream(url)
  File "c:\Users\dsrus\Desktop\School\464\Visual Studio\Hello\Hello.py", line 33, in get_stream        
    m3u8_obj = m3u8.load(stream_url.args['url'])
AttributeError: 'UHSStream' object has no attribute 'args'

I can get twitch, youtube working, not ustream. Any thoughts on this?

my code is the straight up one i grabbed from somewhere

`import time
from datetime import datetime, timedelta, timezone
import urllib
import m3u8
import streamlink
import cv2 #openCV

    def get_stream(url):

    """
    Get upload chunk url
    input: youtube URL
    output: m3u8 object segment
    """
    #Try this line tries number of times, if it doesn't work, 
    # then show the exception on the last attempt
    # Credit, theherk, https://stackoverflow.com/questions/2083987/how-to-retry-after-exception
    tries = 1
    for i in range(tries):
     try:
        streams = streamlink.streams(url)
     except:
        if i < tries - 1: # i is zero indexed
            print(f"Attempt {i+1} of {tries}")
            time.sleep(0.1) #Wait half a second, avoid overload
            continue
        else:
            raise
     break

    stream_url = streams["best"] #Alternate, use '360p'

    m3u8_obj = m3u8.load(stream_url.args['url'])
    return m3u8_obj.segments[0] #Parsed stream

def dl_stream(url, filename, chunks):
    """
    Download each chunk to file
    input: url, filename, and number of chunks (int)
    output: saves file at filename location
    returns none.
    """
    pre_time_stamp = datetime(1, 1, 1, 0, 0, tzinfo=timezone.utc)

    
    #Repeat for each chunk
    #Needs to be in chunks because 
    #  1) it's live
    #  2) it won't let you leave the stream open forever
    i=1
    while i <= chunks:
       
        #Open stream
        stream_segment = get_stream(url)
    
        #Get current time on video
        cur_time_stamp = stream_segment.program_date_time
        #Only get next time step, wait if it's not new yet
        if cur_time_stamp <= pre_time_stamp:
            #Don't increment counter until we have a new chunk
            print("NO   pre: ",pre_time_stamp, "curr:",cur_time_stamp)
            time.sleep(0.5) #Wait half a sec
            pass
        else:
            print("YES: pre: ",pre_time_stamp, "curr:",cur_time_stamp)
            print(f'#{i} at time {cur_time_stamp}')
            #Open file for writing stream
            file = open(filename, 'ab+') #ab+ means keep adding to file
            #Write stream to file
            with urllib.request.urlopen(stream_segment.uri) as response:
                html = response.read()
                file.write(html)
            
            #Update time stamp
            pre_time_stamp = cur_time_stamp
            time.sleep(stream_segment.duration) #Wait duration time - 1

            i += 1 #only increment if we got a new chunk

    return None

def openCVProcessing(saved_video_file):
    '''View saved video with openCV
    Add your other steps here'''
    capture = cv2.VideoCapture(saved_video_file)

    while capture.isOpened():
        grabbed, frame = capture.read()  #read in single frame
        if grabbed == False:
            break

        #openCV processing goes here
        #
        
        cv2.imshow('frame',frame)  #Show the frame
        
        #Shown in a new window, To exit, push q on the keyboard
        if cv2.waitKey(20) & 0xFF == ord('q'):
            break

    capture.release()
    cv2.destroyAllWindows()  #close the windows automatically
                  
tempFile = "temp21.ts"  #files are format ts, open cv can view them
videoURL = "https://ustream.tv/embed/17074538"
#videoURL = "https://www.twitch.tv/nyanners"
#videoURL = "https://video.ibm.com/channel/live-iss-stream"
#https://www.ustream.tv/embed/9408562
dl_stream(videoURL, tempFile, 3)
openCVProcessing(tempFile)``

**Comment by back-to** [comment, marked as answer]
> I can get twitch, youtube working, not ustream. Any thoughts on this?

because you **need Streamlink** for Ustream, there are no static URLs.

> my code is the straight up one i grabbed from somewhere

well it can't work with Ustream.

>  AttributeError: 'UHSStream' object has no attribute 'args'

there is no args here, you can't do it.

```
>>> stream_url.__dict__
{
  'session': <streamlink.session.Streamlink>,
  'poller': <APIPoller(Thread-7, initial daemon)>,
  'first_chunk_data': ChunkData(chunk_id=16, chunk_time=5000,
  hashes={'16': 'a', '10': '6'},
  current_timestamp=datetime.datetime(..., tzinfo=datetime.timezone.utc)),
  'template_url': 'https://uhsakamai-a.akamaihd.net/live/.../.../plain/uhs/1/2/chunk_%s_%s.flv'
}
```

**Reply by dsrushton** [reply]
Thx for the help, i figured out another way!
